### DISCUSSION TITLE
AWS Amplify - Runtime config

### DISCUSSION THREAD

**Discussion by cipami** [nuxt/nuxt#24635] [Questions]
Anyone managed to inject .env to Node runtime on AWS Amplify? 
Only docs about it I found is this: https://docs.aws.amazon.com/en_us/amplify/latest/userguide/ssr-environment-variables.html which doesn't work as Nuxt does not auto load .env files on prod.

**Comment by HaidarZ** [comment]
@cipami have you managed to make it work ?

**Reply by cipami** [reply]
@HaidarZ Sadly not still. any chance you maybe did? 😅

**Reply by adamkutak** [reply]
I'm also having this issue. I just want to use my environment variables on the server routes in Amplify!

**Comment by pi0** [comment]
Hi. Please consider [linked example for Next.js](https://docs.aws.amazon.com/en_us/amplify/latest/userguide/ssr-environment-variables.html) assumes the env variables will get **bundled during build process**. You can do the same by passing build time runtime config.

I highly recommend using platform native Environment Variables for dynamic environment config (otherwise, you essentially need to commit a new `.env` + new deploy to make changes. It is available under [App settings: Environment variables](https://eu-west-1.console.aws.amazon.com/amplify/home?region=eu-west-1#/d3e69m298g51pp/settings/variables) in your dashboard

Docs: https://docs.aws.amazon.com/amplify/latest/userguide/environment-variables.html

<img width="1315" alt="image" src="https://github.com/nuxt/nuxt/assets/5158436/d693f2f4-c93d-47b0-b72d-fbeb84772a9b">

**Reply by HaidarZ** [reply]
So, my build settings will look something like this?

```json
version: 1
frontend:
  phases:
    preBuild:
      commands:
        - nvm use 18
        - corepack enable
        - npx --yes nypm i
    build:
      commands:
        - env >> .env
        - yarn run build
  artifacts:
    baseDirectory: .amplify-hosting
    files:
      - '**/*'
```

**Reply by HaidarZ** [reply]
It didn't work :/

**Reply by pi0** [reply]
(as stated above, doing so will **only** support them for **build time**, and being **bundled** in runtime (if defined in config and prefixed with NUIXT_, it won't give you runtime environment variable configuration)

Also I think both Nitro and Nuxt CLIs don't load `.env` in build time (have to double check), best you can do is to use `source .env` as one additional command, but I wouldn't do this, Amplify has a built-in method to provide environment variables.

**Reply by HaidarZ** [reply]
So, what do you suggest ? for pm2 I was using the following config to load the environment variables at runtime:
```json
module.exports = {
  apps: [
    {
      name: 'MyApp',
      port: '3000',
      exec_mode: 'cluster',
      instances: 'max',
      script: './.output/server/index.mjs',
      node_args: '-r dotenv/config',
    },
  ],
}
```

**Reply by pi0** [reply]
When environment variables are configured [in amplify dashboard](https://docs.aws.amazon.com/amplify/latest/userguide/environment-variables.html) they are automatically available at build-time. The ones that are specified in `nuxt.config`'s `runtimeConfig` section will be bundled during build-time too.

If you want to commit a `.env` file and load it similar to pm2, you might try `NODE_OPTIONS="-r dotenv/config"` or with newer Node.js versions, `NODE_OPTIONS="--env-file .env"` ([read more](https://nodejs.org/en/blog/release/v20.6.0))

**Reply by HaidarZ** [reply]
`--env-file` is a perfect suggestion. Just as a reference to the Nuxt [docs](https://nuxt.com/docs/guide/directory-structure/env#production-preview) , in production , environment variables should be passed to the running engine whatever is the method. runtimeConfig is hydrated ,not bundled. 

References:  
[https://nuxt.com/docs/getting-started/configuration#runtimeconfig-vs-appconfig]( https://nuxt.com/docs/getting-started/configuration#runtimeconfig-vs-appconfig)

**Reply by pi0** [reply]
runtime config is hydrated **and** bundled from build-time env/config.

`.env` support is mainly for Development and Node.js (self) hosting as a convenient way to pass the actual OS environment variables.

For platform deployments, I still highly recommend you to consider their native methods of providing env.

**Reply by HaidarZ** [reply]
If it's really bundled,and knowing that .env file is available during build-time, shouldn't it be available at runtime regardless if the .env file is read or not?

For example, I've built the app on my development machine having a certain .env configuration different than the production one. Then I've uploaded it to the production server, upthere, it will read the runtimeConfig from the .env there. If it was bundled, it should have loaded the development .env file that I've used to build. And if I didn't load any .env file at all for the node production run process, the default values for nuxt config runtimeConfig will be read, not the development .env file I used to build.

**Reply by pi0** [reply]
> If it's really bundled,and knowing that .env file is available during build-time, shouldn't it be available at runtime regardless if the .env file is read or not?

The process is as follows:

**Build step in CI:** 
(platform env injected) > (`.env` manually loaded) > `nuxt build` checks `process.env` and assign defaults > nitro build bundles the runtime config values

**Runtime:** platform starts Node.js process without `.env` > nitro tries to apply env overrides to runtime config if set in `process.env` (which by default is not there, therefore build-time values used)

---

> production server, upthere, it will read the runtimeConfig from the .env there

When production server is your's that's true. it is not case for Amplify Hosting. 

>  If it was bundled, it should have loaded the development .env file that I've used to build

I am not sure to understand this part of your use with AWS Amplify hosting. Usually, you make code changes and commit them to your git repository (`.env` is not committed). Then Amplify's CI/CD will get notified and clones your git repository  and building nuxt/nitro app (without your local `.env`)
 in their environment. Only thing available in Amplify's CI/CD is the environment variables defined in dashboard.

Are you doing it differently by building locally and manually using CLI to directly upload to Amplify hosting?

**Reply by cipami** [reply]
i still have no idea how to make it work, I always end up with the nuxt.config defaults. Considering I have in my project
```
runtimeConfig:{
    public:{
      test: 'NOT SET'
    }
  }
``` 
and I put this to my amplify environment tab `NUXT_PUBLIC_TEST=CORRECT`

What would be a correct way to set it at build time or runtime (doesn't meter since it will always be changeable only via new build)?

**Reply by pi0** [reply]
Something like this should allow build-time envs to be bundled

`test: process.env. NUXT_PUBLIC_TEST || 'NOT SET'`

**Reply by adamkutak** [reply]
I'm having this issue as well. On AWS Amplify, when my Nuxt server gets an env variable with process.env, the variable is not there (only the AWS-added env variables are present). @pi0 how can I access the .env variables in my Nuxt server routes on AWS amplify? The guide for next.js did not work for me.  

Note that I do not have a runtimeconfig. That works fine on Vercel however, but AWS amplify does not work.

**Reply by keithrz** [reply]
> Something like this should allow build-time envs to be bundled
> 
> `test: process.env. NUXT_PUBLIC_TEST || 'NOT SET'`

~~This did not work for me.~~  THIS WORKED!  (I said it didn't at first, but I hadn't realized that my build failed for a completely different reason).

Here is my code, for those who are looking for a more complete example:
```
export default defineNuxtConfig({
  devtools: { enabled: true },
  modules: ["@nuxt/ui"],
  experimental: {
    externalVue: false,
  },
  // need to pre-populate these values with their process.env.* equivalent
  // this is needed for the deployments to staging and prod envs
  // more info here: https://github.com/nuxt/nuxt/discussions/24635
  runtimeConfig: {
    public: {
      aws: {
        appsync: {
          endpoint: process.env.NUXT_PUBLIC_AWS_APPSYNC_ENDPOINT || '',
          region: process.env.NUXT_PUBLIC_AWS_APPSYNC_REGION || '',
        },
        cognito: {
          region: process.env.NUXT_PUBLIC_AWS_COGNITO_REGION || '',
          identityPoolId: process.env.NUXT_PUBLIC_AWS_COGNITO_IDENTITY_POOL_ID || '',
          userPoolId: process.env.NUXT_PUBLIC_AWS_COGNITO_USER_POOL_ID || '',
          clientId: process.env.NUXT_PUBLIC_AWS_COGNITO_CLIENT_ID || '',
        }
      },
    },
  },
})
```

I'm using an .env file for local development, but for my deployments I am setting env vars in AWS Amplify "Environment variables" section as described by @pi0.

~~Does anyone have an example of this working with Nuxt 3 & Amplify?~~ nvm, it works now.

Sidenote here: I feel like I shouldn't need these specific variables defined in this way - Amplify generates the files like `.aws-exports` and `.amplifyconfiguration`, that contain the values I already need, so in theory I could reference one of those files in my code directly, and generate the file using `amplify pull -y --appId MYAPPID --envName $USER_BRANCH`.  But I couldn't get headless amplify pull working for me (even in the Amplify frontend build) because it needed me to authenticate to Amplify Studio.

**Reply by chuckntaylor** [reply]
Wow, glad I stumbled onto this. The documentation does not mention specifically referring to `process.env` when setting the runtimeConfig properties. The docs simply mention using env variables that start with `NUXT_PUBLIC_` if they are going to be used in the `runtimeConfig.public` values.

However, that really is the missing ingredient -especially since the build process forces `NODE_ENV=production`. I was really stumped as to how to setup some other values for my development branch.

As a side note, when setting the runtimeConfig like this, it is not necessary to use the `NUXT_PUBLIC_` prefix for the env variables.

The following worked just fine for me:
```ts
export default defineNuxtConfig({
runtimeConfig: {
    public: {
      appEnv: process.env.APP_ENV || '',
      apiBase: process.env.API_BASE || '',
      cdnBase: process.env.CDN_BASE || '',
      stripePublishableKey: process.env.STRIPE_PUBLISHABLE_KEY || '',
    }
  },
}
```

![EnvVariablesInAmplify](https://github.com/nuxt/nuxt/assets/17050325/ea6d0cb0-105d-45a2-952c-46555fbb4fa9)

**Reply by keithrz** [reply]
> Sidenote here: I feel like I shouldn't need these specific variables defined in this way - Amplify generates the files like `.aws-exports` and `.amplifyconfiguration`, that contain the values I already need, so in theory I could reference one of those files in my code directly, and generate the file using `amplify pull -y --appId MYAPPID --envName $USER_BRANCH`. But I couldn't get headless amplify pull working for me (even in the Amplify frontend build) because it needed me to authenticate to Amplify Studio.

I figured out how to do this.  I needed to remove the "backend" section to my amplify.yml config file, and uncheck the option in Amplify Console:
> Enable full-stack continuous deployments (CI/CD)
Full-stack CI/CD allows you to continously deploy frontend and backend changes on every code commit

Once I did this, the "Backend" part of my deploy generated an aws-exports file which I could use in my imports (and that's the only thing that was done in the "Backend" deploy).  The aws-exports file was the same file generated in my local env after running amplify pull, so the transition from local to deployed env was pretty seamless.

**Reply by adamkutak** [reply]
Note that I used runtimeConfig as well like the above commenters and now it works perfectly fine!

**Comment by ghost** [comment]
Has anyone got the server environment variables to work? I can get them on my client using the public runtimeConfig in my nuxt.config.js, but the server environment variables are always undefined, whether calling them from process.env or using the runtime config. I've used the build commands to inject them into .env and .env.production, but both have the same result.

**Reply by ChristianUlbrich** [reply]
See my post for a technical explanation.

**Comment by benlavalley** [comment]
It is critical you have NUXT_ in your env var naming convention for all vars, public and private. Do this and AWS Amplify deploys should pull them correctly.

**Reply by ChristianUlbrich** [reply]
This is wrong. `NUXT_` cannot be used to override runtimeConfig for deployed AWS Amplify apps. If you want them to apply to the build, you **must** reference them _excplicitly_, see my explanation.

**Reply by benlavalley** [reply]
Yet this is exactly what I had to do to get my app in AWS Amplify to deploy. It's now been working perfectly fine for many months. 

Not sure what else I can say other than the fact that something is being lost in translation here.

My deploys did not work until NUXT_ was added as a prefix to my environment variables referenced in runtimeConfig and defined in my AWS Amplify App's Configure Service -> Runtime envronment variables. 

[keithrz](https://github.com/keithrz) referenced the same thing above back in April.

Zero other changes made - as soon as I updated my runTimeConfig to expect process.env.NUXT_SOMEVAR and NUXT_ to my AWS env vars, my AWS build>deploys from Github commits worked just fine.

If anyone would like more details on my live app running perfectly fine in AWS Amplify using environment variables feel free to message me on the Nuxt discord (ben305).

**Comment by ChristianUlbrich** [comment]
I stumbled on this problem myself and I think I found out, what is happening. It boils down to _build time_ and _runtime_ and people's expectations about, how `environment variables` should be "seen".

I am far from saying I know, what people expect, but I think the expectations are (at least mine were):
* _any_ `process.env` related stuff is seen during _build time_ and _run time_, regardless, where the build happens and regardless, where your Nuxt app is running.
* special `NUXT_` variables should _always_ work.

Turns out, that this is not the case, neither for using Nuxt itself, nor for running Nuxt apps on AWS Amplify. My observations are:

* During the build, _any_ `process.env` in the _runtimeConfig_ is _resolved_ and _baked_ into the generated output
* _AWS Amplify Environment variables_ are **only** exposed to the _build_ in Amplify, **they are not** exposed to the AWS Lambdas actually being run, i.e. this means, you cannot (easily) expose any `process.env` stuff to you server-run code.
  * this means, **you cannot** add any `NUXT_` named _AWS Amplify environment variable_ to _override_ any _Nuxt runtimeConfig_   
* While specially named `NUXT_` _environment variables_, would allow you to override any _runtimeConfig configuration_ (if they get passed), they only allow you to do so on _run time_, **not on build time** (which actually makes sense).
  * this means, you cannot override any _Nuxt runtimeConfig_ for the build with them

Having understood this, for AWS Amplify, this means:
* you must explicitly add any `AWS Amplify Environment variables` you want to use in your Nuxt app to the _runtimeConfig_ and have them explicitly refer to them. They will get baked in, during the _build_ on AWS Amplify:
```
runtimeConfig: {
    // we explicitly pass a NUXT_ env variable, because
    // AWS Amplifiy cannot provide environment variables on runtime, just on build time
    brevoApiKey: process.env.NUXT_BREVO_API_KEY || 'missing',
  },
```  
* changing the _AWS Amplify Environment variables_ will no effect on your deployed app; if you want to have them applied, you need to _build_ again.

While this effectively means, that you might not have to use the `NUXT_` prefix, because you are explicitly referencing them anyway, it has still the effect, that you can still _locally_ (or for any _hosting environment_, that properly supports setting _environment variables_ for _run time_) change the _runtimeConfig_.

_Bonus:_
* `.env` are used for _development_, **and** for the build, but the generated _node server_ does not use it.

**Reply by joeldavuk** [reply]
Not sure if this relates to gen1 or 2, but using 2 the above did not work, using NUXT prefix, and populating runtimeConfig, and using the env section of the build in amplify, had no effect during runtime. It did work with a local build though e.g
`NUXT_DATABASE_URL=your-connection-string node .output/server/index.mjs`

 the only way to make this work on amplify was using the nuxt way of doing it with useRuntimeConfig helper
```
  const config = useRuntimeConfig();
  config.databaseUrl;
```
 and in nuxt.config
 
```
   runtimeConfig: {
    databaseUrl: process.env.NUXT_DATABASE_URL || "",
```
